core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java

```diff
@@ -723,17 +728,22 @@ public V put(K key, V value, long lifespan, TimeUnit unit) {
    }
 
    protected V getAndPutInternal(K key, V value, Metadata metadata) {
+      CacheEntry<K, V> oldEntry = getAndPutInternalEntry(key, value, metadata);
+      return oldEntry != null ? oldEntry.getValue() : null;
+   }
+
+   private CacheEntry<K, V> getAndPutInternalEntry(K key, V value, Metadata metadata) {
       Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED);
       Objects.requireNonNull(value, NULL_VALUES_NOT_SUPPORTED);
-      ValueAndMetadata<V> oldRef = new ValueAndMetadata<>();
+      ByRef<CacheEntry<K, V>> oldEntryRef = new ByRef<>(null);
       boolean hasListeners = this.hasListeners;
       getDataContainer().compute(key, (k, oldEntry, factory) -> {
          if (isNull(oldEntry)) {
             if (hasListeners) {
                CompletionStages.join(cacheNotifier.notifyCacheEntryCreated(key, value, metadata, true, ImmutableContext.INSTANCE, null));
             }
          } else {
-            oldRef.set(oldEntry.getValue(), oldEntry.getMetadata());
+            oldEntryRef.set(oldEntry.clone());
```

Comment:
Maybe just add a comment saying that we have to clone due to possibly updating the value reference below?